File: CalendarWithObjC/CalendarWithObjC/CustomCalendar/Calendar/DateGenerator/DateGenerator.m
//
//  DateGenerator.m
//  CathayBank
//
//  Created by Marvin iOS on 2018/9/12.
//  Copyright © 2018年 YuChen Hsu. All rights reserved.
//

#import "DateGenerator.h"

@implementation DateGenerator

+ (NSDate *)getDateFromInt:(NSInteger)year month:(NSInteger)month day:(NSInteger)day {
    
    if (!year || !month || !day) {
        return nil;
    }
    
    NSCalendar *calendar = [NSCalendar currentCalendar];
    
    NSDateComponents *components = [NSDateComponents new];
    
    components.year = year;
    components.month = month;
    components.day = day;
    components.calendar = calendar;
    
    if (![components isValidDate]) {
        return nil;
    }
    
    NSDate *returnDate = [calendar dateFromComponents:components];
    return returnDate;
}

+ (NSDateComponents *)getDateComponentsFrom:(NSDate *)date {
    
    if (!date) {
        return nil;
    }
    
    NSDateComponents *components = [[NSCalendar currentCalendar] components:NSCalendarUnitDay | NSCalendarUnitMonth | NSCalendarUnitYear fromDate:date];
    
    return components;
    
}

+ (NSArray *)getMonthsNameArray {
    
    NSDateFormatter *dateFormatter = [NSDateFormatter new];
    [dateFormatter setLocale:[[NSLocale alloc] initWithLocaleIdentifier:@"en_US"]];
   
    return [dateFormatter shortMonthSymbols];
}

+ (NSString*)getYearString:(NSDate*)date {
    
    if (!date) {
        return nil;
    }
    
    NSDateFormatter *formatter = [NSDateFormatter new];
    [formatter setDateFormat:@"yyyy"];
    
    return [formatter stringFromDate:date];
}

+ (NSString*)getMonthString:(NSDate *)date {
    
    if (!date) {
        return nil;
    }
    
    NSDateFormatter *formatter = [NSDateFormatter new];
    [formatter setDateFormat:@"MM"];
    
    return [formatter stringFromDate:date];
    
}

+ (NSString*)getDayString:(NSDate*)date {
    
    if (!date) {
        return nil;
    }
    
    NSDateFormatter *formatter = [NSDateFormatter new];
    [formatter setDateFormat:@"dd"];
    
    return [formatter stringFromDate:date];
}

@end

File: CalendarWithSwift/CalendarWithSwift/CustomCalendar/Calendar/DateGenerator/DateGenerator.swift
//
//  DateGenerator.swift
//  CalendarWithSwift
//
//  Created by YuChen Hsu on 2019/7/22.
//  Copyright © 2019 YuChen Hsu. All rights reserved.
//

import Foundation

struct DateGenerator {
    
    static func getDateFrom(year:Int, month:Int, day: Int) -> Date! {
        let calendar = Calendar.current
        
        let dateComponents = DateComponents(calendar: calendar, year: year, month: month, day: day)
        
        if !dateComponents.isValidDate {
            return Date()
        }
        
        let returnDate = dateComponents.date!
        
        return returnDate
    }
    
    static func getDateComponentsFrom(date:Date) -> DateComponents {
        
        let units: Set<Calendar.Component> = [.day, .month, .year]
        let components = Calendar.current.dateComponents(units, from: date)
        
        return components
    }
    
    static func getMonthsNameArray() -> [String] {
        
        let dateFormatter = DateFormatter()
        dateFormatter.locale = Locale(identifier: "en_US")
        
        return dateFormatter.shortMonthSymbols as [String]
    }
    
    static func getYearString(date:Date) -> String {
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy"
        
        return dateFormatter.string(from: date)
    }
    
    static func getMonthString(date:Date) -> String {
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "MM"
        
        return dateFormatter.string(from: date)
    }
    
    static func getDayString(date:Date) -> String {
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "dd"
        
        return dateFormatter.string(from: date)
    }
}
